Sources/SwiftParser/Expressions.swift

```diff
@@ -319,10 +319,13 @@ extension Parser {
 
       let rhs: RawExprSyntax?
       if colon.isMissing {
-        // If the colon is missing there's not much more structure we can
-        // expect out of this expression sequence. Emit a missing expression
-        // to end the parsing here.
-        rhs = RawExprSyntax(RawMissingExprSyntax(arena: self.arena))
+        if let previousTokenKind = currentToken.cursor.previousTokenKind,
+          self.at(TokenSpec(previousTokenKind))
```

Comment:
This will only catch cases where the two tokens that should be separated by the colon are of the same kind. It won’t catch cases like `foo ? 1 someVar`. 

But actually, I’m thinking there’s no reason to ever set `rhs` to a `MissingExprSyntax` (maybe there never was, I don’t know) and just setting it to `nil` seems like the cleaner fix to me.

Reply:
Oh, right!
I just checked the behavior of the Swift Compliler in Xcode and it doesn't seem to check for expressions that should be after the colon when there is no colon.
<img width="50%" alt="스크린샷 2023-05-20 오후 2 05 15" src="https://github.com/apple/swift-syntax/assets/50410213/0014cc35-4547-4f31-b1d5-6cc2b202b934"><img width="50%" alt="스크린샷 2023-05-20 오후 2 11 06" src="https://github.com/apple/swift-syntax/assets/50410213/2f17cd74-c83b-4f6b-a0b0-badb2667f288">

One difference is that the Xcode Swift Compiler does not suggest adding an expression in the following cases 🤔 
<img width="50%" alt="스크린샷 2023-05-20 오후 2 19 21" src="https://github.com/apple/swift-syntax/assets/50410213/c8d61eec-7bc3-4cff-9b7f-2ce412f864f1">
I think the behavior of SwiftSyntax is more accurate in this case, so I'll keep it.

Reply:
In addition, in the case below, we need to check the type of both expressions of the colon, which is not possible with just the tokenKind check.
<img width="50%" alt="스크린샷 2023-05-20 오후 2 38 44" src="https://github.com/apple/swift-syntax/assets/50410213/ae008008-c4c2-4c47-b722-5e4c11f99a0e">
Is there an already implemented function to check for type mismatches?

Reply:
Oh, I now see why you checked for the token kind. 

Let me give you some compiler background: Compilers are usually divided in multiple stages. The interesting stages here are lexing, parsing, name lookup and type checking. 

The purpose of lexing is to take the contents of a file and form tokens from it. For example, it knows that two words separated by a space are two different identifiers instead of one. It also takes care of classifying comments. These are the files in SwiftParser/Lexer

Parsing than takes this stream of tokens and forms a tree from them. And if it can't form a tree from the tokens, the parser will throw an error. Parsing is still what we call syntactical, that is, it doesn't have a clue about types or which declarations a variable, function, etc. references. This is the part of the code base you are working on. 

The next step is name lookup. Name lookup determines the declarations a variable/function/… reference might refer to. This stage can, for example, already take into account when one variable shadows another. Name lookup still doesn't concern itself with type information. If a function is overloaded, for example, it just returns an overload set of functions that might be applicable.

And after that there's type checking, which takes the structure from parsing and the information from type checking to check that all the types match. Because of Swift's rules that you can overload functions based on their return types and other type inference features, this is a whole lot more complicated beast. 

Having said that, name lookup and type checking are not implemented in the swift-syntax repo yet, so there is no way you can check that the types of the arguments match.

Reply:
Ah, I see! Now it all makes sense.
I sincerely appreciate your detailed explanation. 
Thanks to you, I was able to understand it easily 🙇‍♂️